### PR TITLE
Add suport for prometheus 2.1

### PR DESCRIPTION
**Pull request by ebabani**
Most changes are in the configmap due to the new rules format and the
different method of setting the alertmanager target.

Migration notes: https://prometheus.io/docs/prometheus/latest/migration/

### CODE REVIEW

**Comment by kenden** [comment]
Fixes https://github.com/kubernetes/charts/issues/2712

**Comment by niels-s** [comment]
Is there someone who has time to have a look and maybe could approve this? Maybe @viglesiasce or @prydonius?

**Comment by yagonobre** [comment]
the prometheus 2.0 docker image use the user `nobody` then the prometheus can't write to `/data`

**Comment by thuandt** [comment]
I think we just`chmod` every time prometheus `server` started or run as root

**Comment by ebabani** [comment]
@yagonobre That's true but I changed it to run as root.

https://github.com/kubernetes/charts/pull/2767/files#diff-97df733ade0fb9ea384f77bf3a393a0aR28

**Comment by yagonobre** [comment]
@ebabani well done 👏

**Comment by viglesiasce** [comment]
@mgoodness can you take a look?

**Comment by jurgenweber** [comment]
I would of thought you should make a whole new prometheus2 chart... why? So you can do a migration or can you maybe do a migration without it?

Can documentation for the process be added? :)

https://prometheus.io/docs/prometheus/2.0/migration/

Me personally I don't care, I am happy to blow away the data for now.

File: stable/prometheus/templates/server-configmap.yaml [hunk 1]
```diff
@@ -10,5 +10,20 @@ metadata:
...
 data:
-{{ toYaml .Values.serverFiles | indent 2 }}
-{{- end }}
+{{- $alertmanagerHostname := include "prometheus.alertmanager.hostname" . -}}
+{{- $root := . -}}
+{{- range $key, $value := .Values.serverFiles }}
+  {{ $key }}: |
+{{ toYaml $value | default "{}" | indent 4 }}
+{{- if eq $key "prometheus.yml" -}}
+{{- if $alertmanagerHostname -}}
```
Comment (thuandt) [inline stable/prometheus/templates/server-configmap.yaml]: I think we should use kubernetes auto discovery alertmanager in here.

https://prometheus.io/docs/prometheus/2.0/migration/#amsd

```alerting:
  alertmanagers:
  - kubernetes_sd_configs:
      - role: pod
    tls_config:
      ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
    relabel_configs:
    - source_labels: [__meta_kubernetes_pod_label_name]
      regex: alertmanager
      action: keep
    - source_labels: [__meta_kubernetes_namespace]
      regex: {{ .Release.Namespace }}
      action: keep
    - source_labels: [__meta_kubernetes_pod_container_port_number]
      regex:
      action: drop

```

**Review by thuandt — CHANGES_REQUESTED** [review]

**Comment by thuandt** [comment]
This PR working fine. Tested on my cluster.
One thing we can improve is using auto discovery for alertmanager pods

**Comment by ebabani** [comment]
Sorry, I haven't had much time to get back to this recently. I'll post an update with migration instructions and auto discovery for alertmanager pods by next monday/tuesday.

**Comment by vietwow** [comment]
Hi @ebabani ,

How about status of this chart ? Thanks

Best Regards,

File: stable/prometheus/templates/server-configmap.yaml (hunk 1 above)
Comment (thuandt) [reply stable/prometheus/templates/server-configmap.yaml]: btw, `alertmanagers` discovery snippet is extractly like this (tested on my cluster)

```yaml
    alerting:
      alertmanagers:
      - kubernetes_sd_configs:
          - role: pod
        tls_config:
          ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
        bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
        relabel_configs:
        - source_labels: [__meta_kubernetes_namespace]
          regex: {{ .Release.Namespace }} ## maybe not works in configmap, you need assigned to 1 variable first.
          action: keep
        - source_labels: [__meta_kubernetes_pod_label_app]
          regex: prometheus
          action: keep
        - source_labels: [__meta_kubernetes_pod_label_component]
          regex: alertmanager
          action: keep
        - source_labels: [__meta_kubernetes_pod_container_port_number]
          regex:
          action: drop
```

**Comment by ebabani** [comment]
Added auto discovery for alertmanager. Thanks for the help @thuandt 

I also added a section about chart upgrades, and recommend creating a new release if you have an older version of the chart. I verified that the instructions at https://prometheus.io/docs/prometheus/latest/migration/#storage are valid. I deployed a separate release for prometheus 2 and added the old prometheus 1.x release as a remote_read target, allowing me to keep my old data.

**Comment by sfrique** [comment]
I think creating a prometheus2 chart makes sense for a while.

**Comment by StianOvrevage** [comment]
Agree. It's better to just spend a few hours creating a new Prometheus v2 chart that people who are able to use a fresh deploy can use.

And after that spend however many weeks it takes to figure out a way to migrate the old chart to v2 if there is demand for such a migration.

**Comment by ebabani** [comment]
Sounds good to me. I was thinking this morning that it won't be possible to do a clean migration, so a prometheus2 chart makes more sense. This way updates can still be made to the prometheus1 chart.

I can make a new PR over the weekend, unless someone wants to go ahead an make one before then.

**Comment by SuperQ** [comment]
Ping, it would be nice to get this done soon.

File: stable/prometheus/values.yaml [hunk 2]
```diff
@@ -249,7 +249,7 @@ nodeExporter:
   ##
   image:
     repository: prom/node-exporter
-    tag: v0.15.0
+    tag: v0.15.1
```
Comment (SuperQ) [inline stable/prometheus/values.yaml]: `v0.15.2` is available, minor bug fix.

**Comment by lambertjosh** [comment]
I'm concerned about a Prometheus 2.0 specific chart. This seems like a bandaid that would create fragmentation, and additional migrations down the road for users to get back onto the mainline chart.

What about leaving the chart default to be 1.x for now so people aren't surprised on upgrade, and you can manually specify 2.x if you want and deal with no migration? If/when a migration is the available, the default can be bumped.

**Comment by pswenson** [comment]
I think it's acceptable to just upgrade the chart for 2.0 and document a migration path.  I think the default should be 2.0 though... all new projects would want to use 2.0. 

The old 1.x chart is still available for those who don't want to upgrade.

**Comment by ebabani** [comment]
I'll put a larger migration section and a sample `prometheus.yml` for the old release.

Yeah, thinking more about it 2 charts is not the right thing to do long term since some updates would have to be made to both, and features would diverge.

**Review by SuperQ — APPROVED** [review]
LGTM

**Comment by SuperQ** [comment]
@niels-s @mgoodness Can we get this merged soonish?

**Comment by niels-s** [comment]
@SuperQ I would like to as well but we need to get @mgoodness to get this approved and merged. I already assigned it to @mgoodness but so far no luck. Maybe somebody else from the owners could  help us @viglesiasce, @mattfarina, @lachie83 ?

**Comment by SuperQ** [comment]
@niels-s Would it be possible to have an additional OWNERS file in the Prometheus directory?  Then we could have some Prometheus developers able to review and approve changes.

**Comment by jstoja** [comment]
Maybe this would interest @brian-brazil ?
I tried this PR on our test envs and it worked perfectly. (with the securityContext set).

**Comment by jtslear** [comment]
Some of the default values in the `values.yaml` do not match those that are noted as default in the `README.md`. `kubeStateMetrics.image.tag` is one example.

**Comment by lachie83** [comment]
Thanks for raising this. I'm happy to pick this up. Let me test and get back to you

**Comment by thuandt** [comment]
@ebabani How about upgrading alertmanager to [v0.13.0](https://github.com/prometheus/alertmanager/releases/tag/v0.13.0).
Alertmanager now using `kingpin` like prometheus server

File: stable/prometheus/templates/server-deployment.yaml [hunk 3]
```diff
@@ -25,6 +25,8 @@ spec:
         component: "{{ .Values.server.name }}"
         release: {{ .Release.Name }}
     spec:
+      securityContext:
+        runAsUser: 0
```
Comment (lachie83) [inline stable/prometheus/templates/server-deployment.yaml]: Why is this necessary?

Comment (lachie83) [reply stable/prometheus/templates/server-deployment.yaml]: NM - Just read earlier comments

Comment (lachie83) [reply stable/prometheus/templates/server-deployment.yaml]: I have the following issue on my test cluster

```
  Warning  FailedCreate  3m (x17 over 8m)  replicaset-controller  Error creating: pods is forbidden: pod.Spec.SecurityContext.RunAsUser is forbidden
```

Comment (lachie83) [reply stable/prometheus/templates/server-deployment.yaml]: I would rather remove this. Here's a possible solution
```
      initContainers:
      - name: "init-chown-data"
        image: "busybox"
        command: ["chown", "-R", "65534:65534", "{{ .Values.server.persistentVolume.mountPath }}"]
        volumeMounts:
        - name: storage-volume
          mountPath: {{ .Values.server.persistentVolume.mountPath }}
          subPath: "{{ .Values.server.persistentVolume.subPath }}"
```

Comment (ebabani) [reply stable/prometheus/templates/server-deployment.yaml]: Thanks @lachie83 , I like that a lot better. Didn't realize that runAsUser could be disabled with `SecurityContextDeny`.

I'll make this change later today.

Comment (lachie83) [reply stable/prometheus/templates/server-deployment.yaml]: Thanks. Otherwise I think it looks good. Once we get this merged, let's get prometheus 2.1 out there.

**Comment by ebabani** [comment]
Updated the deployment so prometheus no longer needs to run as root (thanks @lachie83 ) and fixed up the readme.

File: stable/prometheus/values.yaml [hunk 4]
```diff
@@ -191,7 +191,7 @@ kubeStateMetrics:
   ##
   image:
     repository: k8s.gcr.io/kube-state-metrics
-    tag: v1.1.0-rc.0
+    tag: v1.1.0
```
Comment (benmathews) [inline stable/prometheus/values.yaml:194]: Can we bump this to v1.2.0?

File: stable/prometheus/values.yaml [hunk 5]
```diff
@@ -332,13 +332,9 @@ server:
   ##
   image:
     repository: prom/prometheus
-    tag: v1.8.1
+    tag: v2.0.0
```
Comment (thuandt) [inline stable/prometheus/values.yaml]: Prometheus 2.1 has a lots of improvements. Can we bump to 2.1?

https://github.com/prometheus/prometheus/releases/tag/v2.1.0

**Comment by ebabani** [comment]
My concern is that I haven't really tested prometheus 2.1 and kubeStateMetrics 1.2 yet, and this PR has been open for a while.

I would rather they be another PR after this is merged,  especially KubeStateMetrics since that should bump the Kubernetes dependency to 1.8+ due to expecting v1beta1.CronJob resource to exist.

If someone wants to use those versions they can always override them for their release.

**Comment by SuperQ** [comment]
@ebabani There are no changes in Prometheus 2.1 that would make any difference to metrics collection.  Most of the changes are fixes and improvements to the TSDB which are transparent to the targets.

That said, I completely agree that we need to just ship this PR so we can move forward with additional changes.  At this rate, 3.0 will be out before we get this merged. :disappointed:

**Comment by lachie83** [comment]
Merging now! Thanks for all the hard work and time that went into this one!

**Comment by jurgenweber** [comment]
yes!!! I am excited. Thanks team!
